controllers/base_controller.go

```diff
@@ -98,47 +98,103 @@ func (r *BaseReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return ctrl.Result{}, err
 	}
 
-	spiffeId := r.makeSpiffeId(obj)
-	parentId := r.makeParentId(obj)
-	if spiffeId == "" || parentId == "" {
+	myEntry := r.makeEntryForObject(obj)
+
+	if myEntry == nil {
 		// Object does not need an entry. This might be a change, so we need to delete any hanging entries.
+		reqLogger.V(1).Info("Deleting entries for pod that no longer needs an ID", "count", len(matchedEntries))
 		err := r.deleteAllEntries(ctx, reqLogger, matchedEntries)
 		return ctrl.Result{}, err
 	}
 
-	createEntryIfNotExistsResponse, err := r.SpireClient.CreateEntryIfNotExists(ctx, &common.RegistrationEntry{
-		Selectors: r.getSelectors(req.NamespacedName),
-		ParentId:  r.makeParentId(obj),
-		SpiffeId:  spiffeId,
-	})
+	var myEntryId string
 
-	if err != nil {
-		reqLogger.Error(err, "Failed to create or update spire entry")
-		return ctrl.Result{}, err
-	}
-	if !createEntryIfNotExistsResponse.Preexisting {
-		reqLogger.Info("Created new spire entry", "entry", createEntryIfNotExistsResponse.Entry)
+	if len(matchedEntries) == 0 {
+		createEntryIfNotExistsResponse, err := r.SpireClient.CreateEntryIfNotExists(ctx, myEntry)
+		if err != nil {
+			reqLogger.Error(err, "Failed to create or update spire entry")
+			return ctrl.Result{}, err
+		}
+		if !createEntryIfNotExistsResponse.Preexisting {
+			reqLogger.Info("Created new spire entry", "entry", createEntryIfNotExistsResponse.Entry)
+		} else {
+			reqLogger.V(1).Info("Found existing identical spire entry", "entry", createEntryIfNotExistsResponse.Entry)
+		}
+		myEntryId = createEntryIfNotExistsResponse.Entry.EntryId
+	} else {
+		// One of the existing entries might already be just right for me... if not, we can choose one and update it
+		requiresUpdate := true
+		for _, entry := range matchedEntries {
+			if r.entryEquals(myEntry, entry) {
+				reqLogger.V(1).Info("Found existing identical enough spire entry", "entry", entry.EntryId)
+				myEntryId = entry.EntryId
+				requiresUpdate = false
+				break
+			}
+		}
+		if requiresUpdate {
+			reqLogger.V(1).Info("Updating existing spire entry to match desired state", "entry", matchedEntries[0].EntryId)
+			// It's important that if multiple instances are running they all pick the same entry here, otherwise
+			// we could have two instances of the registrar delete each others changes. This can only happen if both are
+			// also working off an up to date cache (otherwise the lagging one will pick up the other change later and correct
+			// the mistake.) If that happens then as long as the list of entries is always in the same order, we can guarantee they
```

Comment:
Are two registrars observing the same sets of changes? Otherwise, even if the choice is deterministic, the set on which the algo is run can be different, leading to I don't know what. Would be nice to add half a sentence.

Reply:
In a sensible configuration, no, because you should run with leader election, but I thought it best to handle this case anyway. They'll both see the same changes (or, at least, they'll both see the same final state), but there's no guarantee they'll be perfectly synchronised in point in time. I think in that case the lagging one might undo work of the one that's ahead, but it'll redo the work as it catches up.